File: 2022/python/03_solution.py
from typing import List


def rucksack_duplicate(line: str) -> str:
    n = len(line)
    first, second = set(line[:n // 2]), set(line[n // 2:])
    return (first & second).pop()


def group_duplicate(lines: List[str]) -> str:
    intersection = set(lines[0]) & set(lines[1]) & set(lines[2])
    intersection.remove('\n')
    return intersection.pop()


def get_priority(item: str) -> int:
    if item >= 'a':
        return ord(item) - ord('a') + 1
    else:
        return ord(item) - 38


if __name__ == '__main__':
    with open('../input/03_input.txt', 'r', encoding='utf-8') as f:
        total1 = sum(
            get_priority(rucksack_duplicate(line.strip())) for line in f)
    print(f'Part 1 - Sum of duplicate item priorities: {total1}')
    with open('../input/03_input.txt', 'r', encoding='utf-8') as f:
        lines = f.readlines()
        total2 = 0
        for i in range(0, len(lines), 3):
            total2 += get_priority(group_duplicate(lines[i:i + 3]))
    print(f'Part 2 - Sum of badge priorities: {total2}')

File: 2022/ruby/03_solution.rb
# frozen_string_literal: true

require 'set'

class Set
  def pop
    el = to_a.sample
    delete(el)
    el
  end
end

def rucksack_duplicate(line)
  n = line.length
  first = Set.new(line[0, n / 2].chars)
  second = Set.new(line[n / 2, n].chars)
  (first & second).pop
end

def group_duplicate(line1, line2, line3)
  first = Set.new(line1.chars)
  second = Set.new(line2.chars)
  third = Set.new(line3.chars)
  (first & second & third).pop
end

def get_priority(item)
  if item >= 'a'
    item.ord - 'a'.ord + 1
  else
    item.ord - 38
  end
end

if __FILE__ == $PROGRAM_NAME
  total1 = 0
  lines = IO.readlines('../input/03_input.txt', chomp: true)
  lines.each { |l| total1 += get_priority(rucksack_duplicate(l)) }
  puts "Part 1 - Sum of duplicate item priorities: #{total1}"
  total2 = 0
  (0...lines.length).step(3).each do |i|
    line1 = lines[i]
    line2 = lines[i + 1]
    line3 = lines[i + 2]
    total2 += get_priority(group_duplicate(line1, line2, line3))
  end
  puts "Part 2 - Sum of badge priorities: #{total2}"
end
